troi/playlist.py

```diff
@@ -1,79 +1,148 @@
             else:
-                rec_name = recording.name
-            print("%-60s %-50s" % (rec_name[:59], artist[:49]))
+                print("playlist: %d" % i)
+
+            for recording in playlist.recordings:
+                if not recording:
+                    print("[invalid Recording]")
+                    continue
+
+                if recording.artist is None or recording.artist.name is None:
+                    artist = ""
+                else:
+                    artist = recording.artist.name
+                if recording.name is None:
+                    rec_name = ""
+                else:
+                    rec_name = recording.name
+                print("%-60s %-50s %s" % (rec_name[:59], artist[:49], str(recording.year or "")))
+            print()
+
+
+    def save(self):
+
+        if not self.playlists:
+            raise PipelineError("Playlists have not been generated yet.")
+
+        for i, playlist in enumerate(self.playlists):
+            filename = playlist.filename or "playlist %03d.jspf" % i
+            with open(filename, "w") as f:
+                f.write(ujson.dumps(_serialize_to_jspf(playlist)))
 
     def launch(self):
 
-        if not self.playlist:
-            raise PipelineError("Playlist has not been generated yet.")
+        if not self.playlists:
+            raise PipelineError("Playlists have not been generated yet.")
 
         op = openpost.OpenPost('https://listenbrainz.org/player', keep_file=True, file_name="playlist.html")
-        op.add_key('listens', self.playlist)
+        op.add_key('listens', ujson.dumps(_serialize_to_jspf(self.playlist)))
         op.send_post()
+
+    def submit(self, token, created_for):
+        """
+            Submit the playlist to ListenBrainz.
+
+            token - the ListenBrainz user token to use to submit this playlist.
+            created_for - the ListenBrainz user name for whom this playlist was created.
+                          the token above must be an Approved Playlist Bot in the ListenBrainz
+                          server, otherwise the subission will fail.
+        """
+
+        if not self.playlists:
+            raise PipelineError("Playlists have not been generated yet.")
+
+        playlist_mbids = []
+        for playlist in self.playlists:
+            print(token)
+            r = requests.post(LISTENBRAINZ_PLAYLIST_CREATE_URL,
+                              json=_serialize_to_jspf(playlist, created_for),
+                              headers={"Authorization": "Token " + str(token)})
+            if r.status_code != 200:
+                raise PipelineError("Cannot post playlist to ListenBrainz: HTTP code %d: %s" %
+                                    (r.status_code, r.json()["error"]))
+
+            try:
+                result = ujson.loads(r.text)
+            except ValueError as err:
+                raise PipelineError("Cannot post playlist to ListenBrainz: " + str(err))
```

Comment:
At some point in time I want to revisit the use of ujson here. I think it some cases it still makes sense to use it for speed, but in something like parsing this response, `r.json()` is probably easier to write, since the response is so small. Not a huge blocker to fix

Reply:
I'll skip this for now, lets chat about this tomorrow.

Reply:
We discussed this in person, we can probably use `r.json()` in troi when we read small bits of data from an API using requests. For the bigger picture of json parsing in the LB listen pipeline we need to look at this in more detail